contracts/R0_SpokePool.sol

```diff
@@ -0,0 +1,183 @@
+
+    /// @notice Stores all proofs verified to prevent replay attacks.
+    mapping(bytes32 => bool) public verifiedProofs;
+
+    // Warning: this variable should _never_ be touched outside of this contract. It is intentionally set to be
+    // private. Leaving it set to true can permanently disable admin calls.
+    bool private _adminCallValidated;
+
+    event VerifiedProof(bytes32 indexed dataHash, address caller);
+
+    error NotHubPoolStore();
+    error NotTarget();
+    error AdminCallAlreadySet();
+    error AdminCallNotValidated();
+    error DelegateCallFailed();
+    error AlreadyReceived();
+    error NotImplemented();
+    error InvalidSteelCommitment();
+
+    // All calls that have admin privileges must be fired from within the receiveL1State method that validates that
+    // the input data was published on L1 by the HubPool. This input data is then executed on this contract.
+    // This modifier sets the adminCallValidated variable so this condition can be checked in _requireAdminSender().
+    modifier validateInternalCalls() {
+        // Make sure adminCallValidated is set to True only once at beginning of the function, which prevents
+        // the function from being re-entered.
+        if (_adminCallValidated) {
+            revert AdminCallAlreadySet();
+        }
+
+        // This sets a variable indicating that we're now inside a validated call.
+        // Note: this is used by other methods to ensure that this call has been validated by this method and is not
+        // spoofed.
+        _adminCallValidated = true;
+
+        _;
+
+        // Reset adminCallValidated to false to disallow admin calls after this method exits.
+        _adminCallValidated = false;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(
+        address _steel,
+        address _verifier,
+        address _hubPoolStore,
+        address _wrappedNativeTokenAddress,
+        uint32 _depositQuoteTimeBuffer,
+        uint32 _fillDeadlineBuffer
+    ) SpokePool(_wrappedNativeTokenAddress, _depositQuoteTimeBuffer, _fillDeadlineBuffer, OFT_FEE_CAP) {
+        verifier = _verifier;
+        steel = _steel;
+        hubPoolStore = _hubPoolStore;
+    }
+
+    function initialize(
+        uint32 _initialDepositId,
+        address _crossDomainAdmin,
+        address _withdrawalRecipient
+    ) public initializer {
+        __SpokePool_init(_initialDepositId, _crossDomainAdmin, _withdrawalRecipient);
+    }
+
+    /**
+     * @notice This can be called by an EOA to relay events that the HubPool emits on L1.
+     * @param journalData The public data written by the guest program
+     * @param seal The encoded cryptographic proof (i.e. SNARK).
+     */
+    function receiveL1State(bytes calldata journalData, bytes calldata seal) external validateInternalCalls {
+        // Decode and validate the journal data
+        Journal memory journal = abi.decode(journalData, (Journal));
```

Comment:
nit: why not just have the Journal struct passed in as a struct rather than encoded bytes?

Reply:
We can do that, though we might get some gas savings if we don't. It'll be slightly easier for the off-chain state to receive as a Journal so i'll change optimistically

Reply:
https://github.com/across-protocol/contracts/pull/916/commits/74685ce8b2b9cdf3bef4faf69069cae5459cdf29